### PR TITLE
Etcd version 3 support for skydns

### PR DESCRIPTION
**Pull request by fbdlampayan**
Aims to provide an parameter way of support for skydns.
Default skydns supports etcd v2.
TTL computation for etcd version 3 will be only by default, until RPC for its support is ready in etcdv3.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -82,6 +82,7 @@ SkyDNS' configuration is stored in etcd as a JSON object under the key
 * `ndots`: how many labels a name should have before we allow forwarding. Default to 2.
 * `systemd`: bind to socket(s) activated by systemd (ignores -addr).
 * `path-prefix`: backend(etcd) path prefix, defaults to skydns (i.e. if it is set to `mydns`, the SkyDNS's configuration object should be stored under the key `/mydns/config`).
+* `etcd-version`: etcd version to be supported by skydns during runtime. Supported versions are etcd version 2 or etcd version 3. Support defaults to 2 (etcd version 2).
```
Comment (miekg) [inline README.md]: This needs to be `etcd_version` (i.e. with a underscore).

Also wouldn't it be weird to store how to reach etcd *in* etcd?

Maybe a `-3` flag that specifies use etcd3?

File: backends/etcd/etcd.go [hunk 2]
```diff
@@ -32,6 +34,13 @@ type Backend struct {
 	inflight *singleflight.Group
 }
 
+type Backendv3 struct {
```
Comment (miekg) [inline backends/etcd/etcd.go]: Might be cleaner to put this in a new file in `backends/etcd3/...`?

File: backends/etcd/etcd.go [hunk 3]
```diff
@@ -98,6 +148,34 @@ func (g *Backend) get(path string, recursive bool) (*etcd.Response, error) {
...
 }
 
+func (g *Backendv3) get(path string, recursive bool) (*etcdv3.GetResponse, error) {
+	resp, err := g.inflight.Do(path, func() (interface{}, error) {
+		if recursive == true {
+			r, e := g.client.Get(g.ctx, path, etcdv3.WithPrefix())
+			if r.Kvs == nil {
+				return nil, fmt.Errorf("ErrorCodeKeyNotFound")
+			}
+			if e != nil {
```
Comment (miekg) [inline backends/etcd/etcd.go]: you should check the `e != nil` first probably?

File: backends/etcd/etcd.go [hunk 4]
```diff
@@ -98,6 +148,34 @@ func (g *Backend) get(path string, recursive bool) (*etcd.Response, error) {
 	return resp.(*etcd.Response), err
 }
 
+func (g *Backendv3) get(path string, recursive bool) (*etcdv3.GetResponse, error) {
+	resp, err := g.inflight.Do(path, func() (interface{}, error) {
+		if recursive == true {
+			r, e := g.client.Get(g.ctx, path, etcdv3.WithPrefix())
+			if r.Kvs == nil {
```
Comment (miekg) [inline backends/etcd/etcd.go]: why is this needed? And the error is quite cryptic.

File: main.go [hunk 5]
```diff
@@ -100,6 +104,8 @@ func init() {
 	flag.IntVar(&config.Ndots, "ndots", intEnv("SKYDNS_NDOTS", server.Ndots), "How many labels a name should have before we allow forwarding")
 
 	flag.StringVar(&msg.PathPrefix, "path-prefix", env("SKYDNS_PATH_PREFIX", "skydns"), "backend(etcd) path prefix, default: skydns")
+
+	flag.IntVar(&config.EtcdVersion, "etcd-version", 2, "etcd version to be supported by skydns during runtime. Supported versions are 2 or 3")
```
Comment (miekg) [inline main.go]: As said above a `-3` flag might be cleaner

File: main.go [hunk 6]
```diff
@@ -213,7 +291,42 @@ func validateHostPort(hostPort string) error {
 	return nil
 }
 
-func newEtcdClient(machines []string, certFile, keyFile, caFile, username, password string) (etcd.KeysAPI, error) {
+func newEtcdV3Client(machines []string, tlsCert, tlsKey, tlsCACert string) (*etcdv3.Client, error) {
```
Comment (miekg) [inline main.go]: Moving some stuff out of this file and into a new will make it slightly cleaner.

File: backends/etcd/etcd.go (hunk 2 above)
Comment (fbdlampayan) [inline backends/etcd/etcd.go]: okay

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skynetservices/skydns] Etcd ve..." ]
>fbdlampayan commented on this pull request.
>
>
>
>> @@ -82,6 +82,7 @@ SkyDNS' configuration is stored in etcd as a JSON object under the key
> * `ndots`: how many labels a name should have before we allow forwarding. Default to 2.
> * `systemd`: bind to socket(s) activated by systemd (ignores -addr).
> * `path-prefix`: backend(etcd) path prefix, defaults to skydns (i.e. if it is set to `mydns`, the SkyDNS's configuration object should be stored under the key `/mydns/config`).
>+* `etcd-version`: etcd version to be supported by skydns during runtime. Supported versions are etcd version 2 or etcd version 3. Support defaults to 2 (etcd version 2).
>
>okay will do!
>
>Clarification though, about the -3 flag. What I've intended was something like:
>"./skydns -etcd-version 3"
>
>So instead I should do it like this?
>"./skydns -3"

`./skydns -3`

But I'm open to other suggestion, like the shortness of this option though.

**Comment by fbdlampayan** [comment]
> `./skydns -3`
> 
> But I'm open to other suggestion, like the shortness of this option though.

hmmm how about -etcdv3_enabled, so it will be:
'./skydns -etcdv3_enabled'

This way it is descriptively saying that "run skydns with etcdv3 support enabled"?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skynetservices/skydns] Etcd ve..." ]
>> `./skydns -3`
>>
>> But I'm open to other suggestion, like the shortness of this option though.
>
>hmmm how about -etcdv3_enabled, so it will be:
>'./skydns -etcdv3_enabled'
>
>This way it is descriptively saying that "run skydns with etcdv3 support enabled"?

Don't like the underscore: `-etcd3` ?

/Miek

--
Miek Gieben

**Comment by fbdlampayan** [comment]
> Don't like the underscore: `-etcd3` ?
> 
> /Miek

hahahahaha okay, better! :D
"-etcd3" it is then!

File: backends/etcd/etcd.go [hunk 7]
```diff
@@ -98,6 +148,34 @@ func (g *Backend) get(path string, recursive bool) (*etcd.Response, error) {
...
+				return nil, fmt.Errorf("ErrorCodeKeyNotFound")
+			}
+			if e != nil {
+				return nil, e
+			}
+			return r, e
+		} else {
```
Comment (miekg) [inline backends/etcd/etcd.go]: drop the `else` here as you are returning above.

File: backends/etcd/etcd.go [hunk 8]
```diff
@@ -163,6 +241,39 @@ Nodes:
...
 }
 
+func (g *Backendv3) loopNodes(kv []*mvccpb.KeyValue, nameParts []string, star bool, bx map[bareService]bool) (sx []msg.Service, err error) {
+	if bx == nil {
+		bx = make(map[bareService]bool)
+	}
+
+	index := 0
+	lengthOfEntries := len(kv)
+	for index < lengthOfEntries {
```
Comment (miekg) [inline backends/etcd/etcd.go]: why not `for i := range kv {` ?

File: backends/etcd/etcd.go [hunk 9]
```diff
@@ -163,6 +241,39 @@ Nodes:
...
+
+		serviceInstance := new(msg.Service)
```
Comment (miekg) [inline backends/etcd/etcd.go]: just `serv := new(msg.Service)`

File: backends/etcd/etcd.go [hunk 10]
```diff
@@ -163,6 +241,39 @@ Nodes:
...
+		bx[b] = true
+		serviceInstance.Key = string(kv[index].Key)
+		//TODO: shouldn't be that another call (LeaseRequest) for TTL
+		serviceInstance.Ttl = g.calculateTtl(kv[index], serviceInstance)
+
+		if serviceInstance.Priority == 0 {
+			serviceInstance.Priority = int(g.config.Priority)
+		}
+
+		sx = append(sx, *serviceInstance)
```
Comment (miekg) [inline backends/etcd/etcd.go]: why is this unconditionally added? Why are you not checking the wildcard parts of the name here? I..e what the loopNodes for version 2 does...

File: main.go [hunk 11]
```diff
@@ -213,7 +291,42 @@ func validateHostPort(hostPort string) error {
...
+		TLS: newTLSConfig(tlsCert, tlsKey, tlsCACert),
+	}
+	cli, err := etcdv3.New(etcdCfg)
+	if err != nil {
+		return nil, err
+	}
+	return cli, nil
+}
+
+func newTLSConfig(tlsCertFile, tlsKeyFile, tlsCACertFile string) *tls.Config {
```
Comment (miekg) [inline main.go]: you can just use newHTTPSTransport and extract the `TLSConfig` from there, no reason to duplicate this here.

**Comment by miekg** [comment]
Fyi: I'm also porting this PR to CoreDNS: https://github.com/miekg/coredns/pull/368 (adds way more, but `middleware/etcd/etcd3.go` and friends has parts of this code.

**Comment by fbdlampayan** [comment]
> Fyi: I'm also porting this PR to CoreDNS: miekg/coredns#368 (adds way more, but middleware/etcd/etcd3.go and friends has parts of this code.

sure no prob.

I've now refactored the implementation with your reviews, it is now ready. :)

**Comment by miekg** [comment]
Looks good. But misses tests. It should be "relatively" easy to re-use the ones for etcd2

**Comment by janosi** [comment]
@fbdlampayan Thank you! Great job! :)
@miekg Thank you for accepting it!

**Comment by fbdlampayan** [comment]
Hi @miekg 
Apologies for the long downtime got occupied with other tasks, I'm now able to get back to this. 
I'll handle the UTs and squash bugs when I see them.

I'll put those in another PR, I'll close this one now.
Thanks!

**Comment by miekg** [comment]
great! thank you
